[Event-timing] refresh timestamps of delayed keyup events (due inline text completion)

francovs opened the pull request:
#### 37e18e1196385ba48009420509791d161a9a360c
<pre>
[Event-timing] refresh timestamps of delayed keyup events (due inline text completion)
<a href="https://bugs.webkit.org/show_bug.cgi?id=301445">https://bugs.webkit.org/show_bug.cgi?id=301445</a>
<a href="https://rdar.apple.com/163284173">rdar://163284173</a>

Reviewed by Wenson Hsieh.

We now add some extra processing to how the start time timestamp of
keyup events is handled: we now guarantee that this timestamp is never
older than the start time timestamp of the latest input event. This is
a workaround for <a href="https://webkit.org/b/301443">https://webkit.org/b/301443</a>,
which causes keyup events to be dispatched late, producing event timing
entries with unreasonably large durations.

* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::initializeEventTimingEntry)
* Source/WebCore/page/LocalDOMWindow.h

Canonical link: <a href="https://commits.webkit.org/302233@main">https://commits.webkit.org/302233@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b7a03fb130b5be3faae7d8b5106f2f8030a97dd9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/128321 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/597 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/39153 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/135712 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/79790 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/130193 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/533 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/471 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/97674 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/65578 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/131269 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/364 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/114955 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78275 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/347 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/33062 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/79000 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/108731 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/33546 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/138168 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/442 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/418 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/106211 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/481 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/111296 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106044 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27033 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/358 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/29846 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/52752 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/493 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/63665 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/392 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/459 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/455 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh reviewed (approved):
Seems like an improvement over the current state of things, but I wonder if we could instead use either the `input` event or composition events for these measurements?

francovs committed:
[Event-timing] refresh timestamps of delayed keyup events (due inline text completion)
https://bugs.webkit.org/show_bug.cgi?id=301445
rdar://163284173

Reviewed by Wenson Hsieh.

We now add some extra processing to how the start time timestamp of
keyup events is handled: we now guarantee that this timestamp is never
older than the start time timestamp of the latest input event. This is
a workaround for https://webkit.org/b/301443,
which causes keyup events to be dispatched late, producing event timing
entries with unreasonably large durations.

* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::initializeEventTimingEntry)
* Source/WebCore/page/LocalDOMWindow.h

Canonical link: https://commits.webkit.org/302233@main
